### PR TITLE
eip1962 - bls12-381

### PR DESCRIPTION
**Pull request by jwasinger**

### CODE REVIEW

**Comment by axic** [comment]
@jwasinger @cdetrio Was this superseded by #44 or this adds a different kind of bls12 benchmark?

**Comment by jwasinger** [comment]
This adds rust (native/wasm) benchmarks compiled from matterlabs' eip1962 code.  #44 adds bls12 benchmarks from websnark.  So this is not superseded by #44 .  It's not done yet so it should stay open for now.

File: wasm-engines/rust-code/eip1962_bls_12_381/src/bench.rs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+const SIZE_F1: usize = 6;
+const SIZE_P_G1: usize = SIZE_F1 * 3;
+const SIZE_P_G2: usize = SIZE_F1 * 6;
+const INPUT_SIZE: usize = SIZE_P_G1 + SIZE_P_G2;
+
+fn parse_g1_point(data: &[u8]) -> CurvePoint<CurveOverFpParameters<'static, U384Repr, PrimeField<U384Repr>>> {
+    let p_x = Fp::from_be_bytes(&BLS12_381_FIELD, &data[0..48], true).expect("invalid point");
+    let p_y = Fp::from_be_bytes(&BLS12_381_FIELD, &data[48..96], true).expect("invalid point");
+
+    println!("fuck {:?}", p_x);
```
Comment (axic) [inline wasm-engines/rust-code/eip1962_bls_12_381/src/bench.rs:20]: Would rather not keep this wording in the repo.

File: wasm-engines/rust-code/eip1962_bls_12_381/src/bench.rs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
...,33u8,189u8,184u8,19u8,224u8,43u8,96u8,82u8,113u8,159u8,96u8,125u8,172u8,211u8,160u8,136u8,39u8,79u8,101u8,89u8,107u8,208u8,208u8,153u8,32u8,182u8,26u8,181u8,218u8,97u8,187u8,220u8,127u8,80u8,73u8,51u8,76u8,241u8,18u8,19u8,148u8,93u8,87u8,229u8,172u8,125u8,5u8,93u8,4u8,43u8,126u8,12u8,229u8,213u8,39u8,114u8,125u8,110u8,17u8,140u8,201u8,205u8,198u8,218u8,46u8,53u8,26u8,173u8,253u8,155u8,170u8,140u8,189u8,211u8,167u8,109u8,66u8,154u8,105u8,81u8,96u8,209u8,44u8,146u8,58u8,201u8,204u8,59u8,172u8,162u8,137u8,225u8,147u8,84u8,134u8,8u8,184u8,40u8,1u8,6u8,6u8,196u8,160u8,46u8,167u8,52u8,204u8,50u8,172u8,210u8,176u8,43u8,194u8,139u8,153u8,203u8,62u8,40u8,126u8,133u8,167u8,99u8,175u8,38u8,116u8,146u8,171u8,87u8,46u8,153u8,171u8,63u8,55u8,13u8,39u8,92u8,236u8,29u8,161u8,170u8,169u8,7u8,95u8,240u8,95u8,121u8,190u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,1u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8,0u8];
```
Comment (axic) [inline wasm-engines/rust-code/eip1962_bls_12_381/src/bench.rs:73]: Ugh, a `from_hex` crate would be nicer.

File: wasm-engines/rust-code/eip1962_bls_12_381/gen_pairing_test.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
+from py_ecc.fields import optimized_bls12_381_FQ12 as FQ12
```
Comment (axic) [inline wasm-engines/rust-code/eip1962_bls_12_381/gen_pairing_test.py:1]: Add `#!/usr/bin/env python3` to the beginning and  set the executable flag.

**Comment by axic** [comment]
@jwasinger so can this be merged?

**Comment by cdetrio** [comment]
#44 also adds similar rust code (`bls12-eip1962-rust-native-two-pairings`), but only to scout-engines, to benchmark rust-native using the same test case input as the wasmsnark.

This adds the rust code to wasm-engines, where in addition to benchmarking the rust-native (using some inputs or test cases from `gen_pairing_test.py`), it would compile the rust to wasm and benchmark the rust-wasm in all the standalone engines. Then rust-wasm could be compared to wasmsnark-wasm.

It's not done yet because the inputs generated with `gen_pairing_test.py` need to go into `wasm-engines/inputvectors`, where the template filler will put them into the rust source code to generate the standalone builds. The inputs should be same as used in scout-engines (I haven't checked them yet).

**Pull request closed without merging**